{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f595afe0-f3a7-4005-99a4-62d5d4e7a68b",
   "metadata": {},
   "source": [
    "# Feature selection RSF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b061ee54-66cf-4863-9799-39b2bda28bb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    second, second<-\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:lubridate’:\n",
      "\n",
      "    %within%\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "setwd(\"/home/data/project_code/landstrom_core/prognostic_model_development_revision/r/notebooks\")\n",
    "\n",
    "library(tidyverse)\n",
    "library(survival)\n",
    "library(survminer)\n",
    "library(glmnet)\n",
    "library(WriteXLS)\n",
    "library(ggfortify)\n",
    "library(circlize)\n",
    "library(ComplexHeatmap)\n",
    "library(parallel)\n",
    "library(broom)\n",
    "library(survcomp)\n",
    "library(survivalROC)\n",
    "library(gtsummary)\n",
    "source(\"../getTCGAData.R\")\n",
    "source(\"../preprocessTCGAData.R\")\n",
    "source(\"../KM_analysis.R\")\n",
    "source(\"../Heatmaps.R\")\n",
    "source(\"../enet.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0836de2b-b91f-442e-a24b-d315d85001a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(out.dir.data, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/data/rdata/manuscript_work/revisions/tgfbeta/PRAD' already exists”\n"
     ]
    }
   ],
   "source": [
    "# Define the cancer type \n",
    "cancer.type = \"PRAD\"\n",
    "\n",
    "# Read in the table including the clinical features for each cancer type\n",
    "clin.feat.tb = read.table(\"/lustre/projects/landstrom_core/data/clin_features_final.csv\", sep = \"\\t\", header = T)\n",
    "\n",
    "# Get Clinical variables\n",
    "clin.var = unlist(strsplit(clin.feat.tb$Features[clin.feat.tb$Ctype == cancer.type], split = \",\"))\n",
    "\n",
    "# Ensembl id mapping file \n",
    "ens.id.mapping = \"/home/organisms/Human/hg38/Homo_sapiens.GRCh38_March2022/ENSEMBLE_to_SYMBOL.csv\"\n",
    "\n",
    "# Output dir \n",
    "out.dir.data = file.path(\"/lustre/projects/landstrom_core/data/rdata/manuscript_work/revisions/tgfbeta\", cancer.type)\n",
    "dir.create(out.dir.data, recursive = T)\n",
    "\n",
    "tcga.cn = getTCGACopyNumberData(cancer.type)\n",
    "\n",
    "tcga.expr = getTCGAExpressionData(cancer.type, annotation.file = ens.id.mapping)\n",
    "\n",
    "# Get cancer specific clinical data \n",
    "tcga.clin = getClinData(cancer.type)\n",
    "\n",
    "# Get the end point related clinical data \n",
    "tcga.endpoints = getClinEndpointData(cancer.type) %>% dplyr::select(bcr_patient_barcode, OS, OS.time, DSS, DSS.time, DFI, DFI.time, PFI, PFI.time)\n",
    "\n",
    "# Merge end point data to clinical data \n",
    "tcga.clin = dplyr::left_join(tcga.clin, tcga.endpoints, by = \"bcr_patient_barcode\")\n",
    "\n",
    "write.csv(tcga.clin, file.path(out.dir.data, \"clinical_data.csv\"))\n",
    "\n",
    "tcga.cn = selectPrimaryT(tcga.cn)\n",
    "\n",
    "tcga.cn = dropDuplicateSamples(tcga.cn)\n",
    "\n",
    "tcga.cn.datamat = prepareDataMatrix(tcga.cn)\n",
    "saveRDS(tcga.cn.datamat, file = file.path(out.dir.data, \"copy_number_status.rds\"))\n",
    "\n",
    "tcga.expr = selectPrimaryT(tcga.expr)\n",
    "\n",
    "tcga.expr = dropDuplicateSamples(tcga.expr)\n",
    "\n",
    "tcga.expr.datamat = prepareDataMatrix(tcga.expr)\n",
    "\n",
    "saveRDS(tcga.expr.datamat, \n",
    "        file = file.path(out.dir.data, \"raw_expressions.rds\"))\n",
    "\n",
    "tcga.dataset = mergeTCGAdata(clin.data = tcga.clin,\n",
    "                                  data = list(\"CN\" = tcga.cn.datamat, \n",
    "                                              \"EXP\" = tcga.expr.datamat), \n",
    "                                  data.suffixes = c(\"cn\",\"exp\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1d4423db-800f-4a5b-a68d-ed055601c83a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define function for adding the clinical variables \n",
    "addClinVar = function(data, clin.var) {\n",
    "    if (\"Age\" %in% clin.var) {\n",
    "        data$CLIN$Age.clin <- data$CLIN$age_at_diagnosis.clin\n",
    "    } \n",
    "    if (\"Tumor.stage\" %in% clin.var){\n",
    "        data$CLIN$Tumor.stage.clin = factor(map_chr(data$CLIN$ajcc_pathologic_stage.clin, reformatTumorStage))\n",
    "    }\n",
    "    if (\"Gender\" %in% clin.var){\n",
    "        data$CLIN$Gender.clin <- factor(data$CLIN$gender.clin)    \n",
    "    } \n",
    "    if (\"Gleason.group\" %in% clin.var) {\n",
    "        \n",
    "        # Determine the Gleason group \n",
    "        data$CLIN$Gleason.group.clin = map2_chr(data$CLIN$primary_gleason_grade.clin, \n",
    "                                           data$CLIN$secondary_gleason_grade.clin, \n",
    "                                           determineGleasonGroup)\n",
    "\n",
    "        # Set up the factor levels \n",
    "        data$CLIN$Gleason.group.clin = factor(data$CLIN$Gleason.group.clin, \n",
    "                                    levels = c(\"Gleason_group_1\", \"Gleason_group_2\"))\n",
    "    }\n",
    "    return(data)\n",
    "}\n",
    "\n",
    "# Add clinical variables to dataset\n",
    "tcga.dataset = addClinVar(tcga.dataset, clin.var)\n",
    "\n",
    "saveRDS(tcga.dataset, file = file.path(out.dir.data, \"tcga.dataset.rds\"))\n",
    "\n",
    "rm(\"tcga.expr\")\n",
    "rm(\"tcga.cn\")\n",
    "rm(\"tcga.clin\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cf222173-08e8-4fdd-b24d-39bdb5b423ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.res.root, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD' already exists”\n",
      "Warning message in dir.create(dir.res.km):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots' already exists”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"APPL2\"  \"APPL1\"  \"AURKB\"  \"AURKA\"  \"VPS4A\"  \"VPS4B\"  \"KIF23\"  \"TRAF6\" \n",
      " [9] \"PTEN\"   \"AR\"     \"TP53\"   \"MYC\"    \"TGFBR1\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'name'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "# Try here converting \"-\" characters to 0 for compatibility \n",
    "colnames(tcga.dataset$EXP) = str_replace(colnames(tcga.dataset$EXP), \"-\", \"0\")\n",
    "\n",
    "# Read in the preprocessed dataset if continued \n",
    "#tcga.dataset = readRDS(file.path(out.dir.data, \"tcga.dataset.rds\"))\n",
    "\n",
    "# Raw expression data \n",
    "#tcga.expr.raw.datamat = readRDS(file.path(out.dir.data, \"raw_expressions.rds\"))\n",
    "\n",
    "# Define and create the root directory for results \n",
    "dir.res.root = file.path(\"/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta/\", cancer.type)\n",
    "dir.create(dir.res.root, recursive = T)\n",
    "\n",
    "# Define and create the results for the KM analysis \n",
    "dir.res.km = file.path(dir.res.root, \"Kaplan_Meier_plots\")\n",
    "dir.create(dir.res.km)\n",
    "\n",
    "tgfbeta.genes.df = read.csv(\"/lustre/projects/landstrom_core/data/TGFbeta_list_test.tsv\", sep = \"\\t\", header = F)\n",
    "\n",
    "# Read in the original gene list \n",
    "customer.genes.df = read.csv(\"/lustre/projects/landstrom_core/data/Customer_genes.tsv\", sep = \"\\t\", header = F)\n",
    "customer.genes = customer.genes.df$V1\n",
    "print(customer.genes)\n",
    "\n",
    "# Modified gene names \"-\" to 0 conversion for compatibility\n",
    "gene.list  = unique(c(tgfbeta.genes.df$V1, customer.genes))\n",
    "\n",
    "clinical.end.point.stats = tcga.dataset$CLIN %>% \n",
    "                                   dplyr::select(c(\"OS.clin\",\"DSS.clin\",\"DFI.clin\",\"PFI.clin\")) %>%\n",
    "                                   pivot_longer(everything()) %>%\n",
    "                                   mutate(value = factor(value)) %>%\n",
    "                                   group_by(name, value) %>%\n",
    "                                   summarise(N = n()) %>% \n",
    "                                   pivot_wider(names_from =  value,\n",
    "                                               values_from = N)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8e63024b-b140-4676-b58e-285ba4b0034d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `complete` -> `complete...1`\n",
      "\u001b[36m•\u001b[39m `complete` -> `complete...2`\n",
      "\u001b[36m•\u001b[39m `complete` -> `complete...3`\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Including 481 cases out of 497 cases\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `complete` -> `complete...1`\n",
      "\u001b[36m•\u001b[39m `complete` -> `complete...2`\n",
      "\u001b[36m•\u001b[39m `complete` -> `complete...3`\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Including 480 cases out of 497 cases\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `complete` -> `complete...1`\n",
      "\u001b[36m•\u001b[39m `complete` -> `complete...2`\n",
      "\u001b[36m•\u001b[39m `complete` -> `complete...3`\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Including 330 cases out of 497 cases\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `complete` -> `complete...1`\n",
      "\u001b[36m•\u001b[39m `complete` -> `complete...2`\n",
      "\u001b[36m•\u001b[39m `complete` -> `complete...3`\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Including 481 cases out of 497 cases\"\n"
     ]
    }
   ],
   "source": [
    "# Here we store all the training and validation splits \n",
    "train_and_validation_ls = list()\n",
    "\n",
    "# Variables selected \n",
    "variables_selected_ls = list()\n",
    "\n",
    "# Number of samples in training and validation cohorts \n",
    "nsamples_step1_ls = list()\n",
    "\n",
    "selectVariables = function(clinical.endpoint,\n",
    "                           clinical.variables,\n",
    "                           gene.list, \n",
    "                           data.suffixes){\n",
    "\n",
    "    # Construct the clinical features   \n",
    "    clinical.features =  c(paste0(clinical.endpoint, \".clin\"),\n",
    "                           paste0(clinical.endpoint, \".time.clin\"),\n",
    "                           paste0(clinical.variables, \".clin\"))\n",
    "\n",
    "    # Constructed the sequencing data features \n",
    "    seq.features = unlist(map(data.suffixes, \n",
    "                           .f = function(x, gene.list){paste(gene.list, x, sep = \".\")}, \n",
    "                           gene.list = gene.list))\n",
    "\n",
    "    # Construct the list of variables \n",
    "    feature.ls = c(clinical.features, seq.features)\n",
    "\n",
    "    return(feature.ls)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "#\n",
    "# Function for splitting data randomly into \n",
    "# training and validation set \n",
    "#\n",
    "splitCases = function(obj, split, variables, seed) {\n",
    "    \n",
    "    \n",
    "  # keep track of statistics \n",
    "  ncases.initial = c()\n",
    "    \n",
    "  complete.ls = list()\n",
    "    \n",
    "  # We will only select the cases with complete data \n",
    "  for (i in 1:length(names(obj))){\n",
    "      \n",
    "      # Extract the data \n",
    "      data.type = names(obj)[i]\n",
    "      data = obj[[data.type]]\n",
    "      \n",
    "      # Find the variables of interest \n",
    "      variables.dat = variables[variables %in% colnames(data)]\n",
    "      \n",
    "      \n",
    "      # Check how many individuals there are with non missing data\n",
    "      # Boolean\n",
    "      complete = complete.cases(data[,variables.dat])\n",
    "      \n",
    "      complete.ls[[i]] = as.data.frame(complete)\n",
    "      \n",
    "      # Update the number of complete cases\n",
    "      ncases.initial = c(ncases.initial, sum(complete))\n",
    "      \n",
    "      \n",
    "  }\n",
    "    \n",
    "  # Select complete cases \n",
    "  complete.across.all = apply(bind_cols(complete.ls),1, all)\n",
    "    \n",
    "  print(paste0(\"Including \", sum(complete.across.all)  ,\" cases out of \", max(ncases.initial), \" cases\"))\n",
    "\n",
    "  \n",
    "  # Get the list of complete samples to be included \n",
    "  samples.included = obj$CLIN$Participant.ID[complete.across.all]\n",
    "  \n",
    "  if (split != 1.0) {\n",
    "    \n",
    "    set.seed(seed)\n",
    "    trainIdx <- sample(length(samples.included), split*length(samples.included), replace = FALSE)\n",
    "    \n",
    "    samples.train <- samples.included[trainIdx]\n",
    "    samples.valid <- samples.included[-trainIdx]\n",
    "       \n",
    "  } else {\n",
    "    \n",
    "    data.train <- samples.included\n",
    "    data.valid <- samples.included\n",
    "    \n",
    "  } \n",
    "  result.split = list(\"train\" = samples.train,\n",
    "                      \"validation\" = samples.valid)\n",
    "  return(result.split)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "for (end.point in c(\"OS\",\"DSS\",\"DFI\",\"PFI\")){\n",
    "\n",
    "    \n",
    "    #\n",
    "    # No need to change the first part \n",
    "    #\n",
    "\n",
    "    #\n",
    "    # MODIFICATION. HERE WE SELECT CLIN ENDPOINT SPECIFIC VARIABLES (DE GENES)\n",
    "    # \n",
    "    #gene.list = gene.lists[[end.point]]\n",
    "    \n",
    "    # Selected variables \n",
    "    variables.selected = selectVariables(clinical.endpoint = end.point,\n",
    "                                         clinical.variables = clin.var,\n",
    "                                         gene.list = gene.list, \n",
    "                                         data.suffixes = c(\"cn\",\"exp\"))\n",
    "    \n",
    "    variables_selected_ls[[end.point]] = variables.selected\n",
    "    \n",
    "    \n",
    "    #\n",
    "    # Splitting function needs to be change to accommondate the \n",
    "    # altered data structure \n",
    "    #\n",
    "    \n",
    "    # Data set is split randomly into training and validation sets. Only complete cases \n",
    "    # are selected.\n",
    "    train_and_validation = splitCases(obj = tcga.dataset, \n",
    "                                  split = 0.75, \n",
    "                                  variables = variables.selected, \n",
    "                                  seed = 42)\n",
    "    \n",
    "    # Update list\n",
    "    train_and_validation_ls[[end.point]] = train_and_validation \n",
    "    \n",
    "    \n",
    "    # Store number of  \n",
    "    nsamples.step1 = c(length(train_and_validation$train), length(train_and_validation$validation))\n",
    "    names(nsamples.step1) = c(\"ntrain.step1\", \"nvalid.step1\")\n",
    "    nsamples_step1_ls[[end.point]] = nsamples.step1\n",
    "}\n",
    "\n",
    "splitDataset = function(obj, train_and_validation_ls){\n",
    "\n",
    "    # \n",
    "    split.obs.by.endpoint = list()\n",
    "    \n",
    "    for (end.point in names(train_and_validation_ls)){\n",
    "        \n",
    "        train.samples = train_and_validation_ls[[end.point]]$train\n",
    "        validation.samples = train_and_validation_ls[[end.point]]$validation\n",
    "        \n",
    "        # New entry \n",
    "        split.obs.by.endpoint[[end.point]] = list()\n",
    "        \n",
    "        for (data.type in names(obj)){\n",
    "            \n",
    "            split.obs.by.endpoint[[end.point]][[\"CLIN\"]] = list()  \n",
    "            split.obs.by.endpoint[[end.point]][[\"CLIN\"]][[\"train\"]] = obj$CLIN %>% \n",
    "                          dplyr::filter(Participant.ID %in% train.samples)\n",
    "            split.obs.by.endpoint[[end.point]][[\"CLIN\"]][[\"validation\"]] = obj$CLIN %>% \n",
    "                          dplyr::filter(Participant.ID %in% validation.samples)\n",
    "                    \n",
    "            split.obs.by.endpoint[[end.point]][[\"EXP\"]] = list()    \n",
    "            split.obs.by.endpoint[[end.point]][[\"EXP\"]][[\"train\"]] = obj$EXP %>% \n",
    "                          dplyr::filter(Participant.ID %in% train.samples)\n",
    "            split.obs.by.endpoint[[end.point]][[\"EXP\"]][[\"validation\"]] = obj$EXP %>% \n",
    "                          dplyr::filter(Participant.ID %in% validation.samples)\n",
    "            \n",
    "            split.obs.by.endpoint[[end.point]][[\"CN\"]] = list()\n",
    "            split.obs.by.endpoint[[end.point]][[\"CN\"]][[\"train\"]] = obj$CN %>% \n",
    "                          dplyr::filter(Participant.ID %in% train.samples)           \n",
    "            split.obs.by.endpoint[[end.point]][[\"CN\"]][[\"validation\"]] = obj$CN %>% \n",
    "                          dplyr::filter(Participant.ID %in% validation.samples)\n",
    "        }\n",
    "    }\n",
    "    return(split.obs.by.endpoint)\n",
    "}\n",
    "\n",
    "tcga.dataset.splitted = splitDataset(tcga.dataset, train_and_validation_ls)\n",
    "\n",
    "convertAge = function(x){\n",
    "    return(x/360)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2b87610b-d60c-4d46-ba11-b88760b25e17",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "variance of dispersion residuals not estimated (necessary only for differential expression calling)\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "variance of dispersion residuals not estimated (necessary only for differential expression calling)\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "variance of dispersion residuals not estimated (necessary only for differential expression calling)\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "variance of dispersion residuals not estimated (necessary only for differential expression calling)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "prepareSummary = function(end.point, data){\n",
    "    # Convert age \n",
    "    data[[end.point]]$CLIN$train$Age.clin = convertAge(data[[end.point]]$CLIN$train$Age.clin)\n",
    "    data[[end.point]]$CLIN$validation$Age.clin = convertAge(data[[end.point]]$CLIN$validation$Age.clin)\n",
    "    \n",
    "    a = data[[end.point]]$CLIN$train %>%\n",
    "          tbl_summary(include = paste0(clin.var, \".clin\")) %>% as_tibble()\n",
    "    \n",
    "    b = data[[end.point]]$CLIN$validation %>%\n",
    "          tbl_summary(include = paste0(clin.var, \".clin\")) %>% as_tibble()\n",
    "    \n",
    "    test = cbind(a,b)\n",
    "    test = test[,-3]\n",
    "    test = t(test)\n",
    "    test = as.data.frame(test)\n",
    "    colnames(test) = test[1,]\n",
    "    test = test[-1,]\n",
    "    test$End.point = end.point\n",
    "    return(test)\n",
    "}\n",
    "\n",
    "clin.summary.table = bind_rows(map(c(\"OS\",\"DSS\",\"PFI\",\"DFI\"), \n",
    "                prepareSummary, \n",
    "                data = tcga.dataset.splitted))\n",
    "\n",
    "for (end.point in names(train_and_validation_ls)){\n",
    "    \n",
    "    # Counts and VST for training data\n",
    "    counts.training = expDataToMatrix(tcga.dataset.splitted[[end.point]]$EXP$train)\n",
    "    \n",
    "    vst.transf.training.obj = performVSTtraining(counts.training)\n",
    "    \n",
    "    # Counts for evaluation \n",
    "    counts.validation = expDataToMatrix(tcga.dataset.splitted[[end.point]]$EXP$validation)\n",
    "    vst.transf.validation.counts = performVSTtest(counts = counts.validation, \n",
    "                                                  disp.function.train = vst.transf.training.obj$disp.function)\n",
    "         \n",
    "    tcga.dataset.splitted[[end.point]]$EXP$train = MatrixToExpdata(vst.transf.training.obj$vst.counts)\n",
    "    tcga.dataset.splitted[[end.point]]$EXP$validation = MatrixToExpdata(vst.transf.validation.counts)\n",
    "    \n",
    "}\n",
    "\n",
    "saveRDS(tcga.dataset.splitted, file.path(out.dir.data, \"tcga.dataset_splitted.rds\"))\n",
    "\n",
    "rm(tcga.dataset)\n",
    "\n",
    "tcga.dataset.splitted = readRDS(file.path(out.dir.data, \"tcga.dataset_splitted.rds\"))\n",
    "\n",
    "tcga.dataset.merged =  mergeDataTypes(tcga.dataset.splitted )\n",
    "\n",
    "saveRDS(tcga.dataset.merged, file.path(out.dir.data, \"tcga.dataset_merged.rds\"))\n",
    "\n",
    "rm(tcga.dataset.splitted)\n",
    "\n",
    "tcga.dataset.merged = readRDS(file.path(out.dir.data, \"tcga.dataset_merged.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "375f3034-1b02-456e-b8b3-d4d2a762326f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store summary statistics \n",
    "summary.stats.ls = list()\n",
    "\n",
    "# Iterate over end points \n",
    "for (end.point in c(\"OS\",\"DSS\",\"DFI\",\"PFI\")){\n",
    "    \n",
    "    exp.summary.training = prepSummaryExp(x = tcga.dataset.merged[[end.point]]$train, \n",
    "                                          raw.data = tcga.expr.datamat,\n",
    "                                          variables = variables.selected, type = \"exp\")\n",
    "\n",
    "    cn.summary.training = prepSummaryCN(tcga.dataset.merged[[end.point]]$train, \n",
    "                                        variables = variables.selected, \n",
    "                                        type = \"cn\")\n",
    "    \n",
    "    summary.stats.ls[[end.point]] = list(\"exp.summary.training\" = exp.summary.training,\n",
    "                                         \"cn.summary.training\" = cn.summary.training)\n",
    "}\n",
    "\n",
    "\n",
    "# Store filtered variables \n",
    "variables.selected.filtered.ls = list()\n",
    "\n",
    "# Iterate over end points \n",
    "for (end.point in c(\"OS\",\"DSS\",\"DFI\",\"PFI\")){\n",
    "\n",
    "    exp.features.keep = summary.stats.ls[[end.point]]$exp.summary.training %>% \n",
    "                          filter(`Median expression` > 20, \n",
    "                                 `Fraction of zero expression` < 0.75)\n",
    "\n",
    "    cn.features.keep = summary.stats.ls[[end.point]]$cn.summary.training %>% \n",
    "                          filter(`Maximum fraction of aberrations` > 0.15) \n",
    "\n",
    "    # Update the summary tables \n",
    "    summary.stats.ls[[end.point]]$exp.summary.training$Selected = ifelse(summary.stats.ls[[end.point]]$exp.summary.training$name %in% exp.features.keep$name, \"Yes\", \"No\")\n",
    "    summary.stats.ls[[end.point]]$cn.summary.training$Selected = ifelse(summary.stats.ls[[end.point]]$cn.summary.training$name %in% cn.features.keep$name, \"Yes\", \"No\")\n",
    "\n",
    "    # Collect the variables into vector \n",
    "    variables.selected.filtered.ls[[end.point]] = filterFeatures(variables_selected_ls[[end.point]], exp.features.keep$name, type = \"exp\")\n",
    "    variables.selected.filtered.ls[[end.point]]= filterFeatures(variables_selected_ls[[end.point]], cn.features.keep$name, type = \"cn\")\n",
    "    \n",
    "}\n",
    "\n",
    "saveRDS(variables.selected.filtered.ls,  \n",
    "        file.path(out.dir.data,\"variables_selected_filtered_ls.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f9e9b09a-5f78-43c0-badf-f5974226c343",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(file.path(dir.res.km, end.point)):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/OS' already exists”\n",
      "Warning message in dir.create(out.exp.train.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/OS/Expression/Training_data' already exists”\n",
      "Warning message in dir.create(out.exp.valid.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/OS/Expression/Validation_data' already exists”\n",
      "Warning message in dir.create(out.cn.train.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/OS/CN/Training_data' already exists”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n",
      "NULL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(out.cn.valid.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/OS/CN/Validation_data' already exists”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n",
      "NULL\n",
      "NULL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(file.path(dir.res.km, end.point)):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/DSS' already exists”\n",
      "Warning message in dir.create(out.exp.train.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/DSS/Expression/Training_data' already exists”\n",
      "Warning message in dir.create(out.exp.valid.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/DSS/Expression/Validation_data' already exists”\n",
      "Warning message in dir.create(out.cn.train.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/DSS/CN/Training_data' already exists”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n",
      "NULL\n",
      "NULL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in pchisq(chi, df, lower.tail = FALSE):\n",
      "“NaNs produced”\n",
      "Warning message in pchisq(chi, df, lower.tail = FALSE):\n",
      "“NaNs produced”\n",
      "Warning message in pchisq(chi, df, lower.tail = FALSE):\n",
      "“NaNs produced”\n",
      "Warning message in pchisq(chi, df, lower.tail = FALSE):\n",
      "“NaNs produced”\n",
      "Warning message in dir.create(out.cn.valid.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/DSS/CN/Validation_data' already exists”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n",
      "NULL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(file.path(dir.res.km, end.point)):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/DFI' already exists”\n",
      "Warning message in dir.create(out.exp.train.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/DFI/Expression/Training_data' already exists”\n",
      "Warning message in dir.create(out.exp.valid.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/DFI/Expression/Validation_data' already exists”\n",
      "Warning message in dir.create(out.cn.train.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/DFI/CN/Training_data' already exists”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n",
      "NULL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(out.cn.valid.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/DFI/CN/Validation_data' already exists”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n",
      "NULL\n",
      "NULL\n",
      "NULL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(file.path(dir.res.km, end.point)):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/PFI' already exists”\n",
      "Warning message in dir.create(out.exp.train.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/PFI/Expression/Training_data' already exists”\n",
      "Warning message in dir.create(out.exp.valid.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/PFI/Expression/Validation_data' already exists”\n",
      "Warning message in dir.create(out.cn.train.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/PFI/CN/Training_data' already exists”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n",
      "NULL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(out.cn.valid.dir, recursive = T):\n",
      "“'/lustre/projects/landstrom_core/results/prognostic_model_development_revised_cell_cycle/tgfbeta//PRAD/Kaplan_Meier_plots/PFI/CN/Validation_data' already exists”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n",
      "NULL\n",
      "NULL\n"
     ]
    }
   ],
   "source": [
    "variables.selected.filtered.ls = readRDS(file.path(out.dir.data,\"variables_selected_filtered_ls.rds\"))\n",
    "\n",
    "# Store the KM tables \n",
    "km.pvalue.table.ls = list()\n",
    "\n",
    "# Store the significant features \n",
    "significant.features.ls = list()\n",
    "\n",
    "# Iterate over end points \n",
    "for (end.point in c(\"OS\",\"DSS\",\"DFI\",\"PFI\")){\n",
    "\n",
    "    # Create dir for plots \n",
    "    dir.create(file.path(dir.res.km, end.point))\n",
    "    \n",
    "    if (nrow(tcga.dataset.merged[[end.point]]$train) > 0){\n",
    "    \n",
    "        # Run univariate KM\n",
    "        km.pvalue.table = runUnivariateKM(input.data = tcga.dataset.merged[[end.point]], \n",
    "                                          variables = variables.selected.filtered.ls[[end.point]],\n",
    "                                          clinical.endpoint = end.point,\n",
    "                                          out.dir = file.path(dir.res.km, end.point),\n",
    "                                          plots = T)\n",
    "    \n",
    "    \n",
    "        # Sort the results based on the training p-value and write the results to output\n",
    "        km.pvalue.table = km.pvalue.table %>% dplyr::arrange(pvalues.training)\n",
    "        km.pvalue.table$Selected = ifelse(km.pvalue.table$pvalues.training < 0.05, \"Yes\", \"No\") \n",
    "        write.csv(km.pvalue.table, file.path(dir.res.km, paste0(end.point, \"_LogRank_pvalues.csv\")))\n",
    "    \n",
    "        km.pvalue.table.ls[[end.point]] = km.pvalue.table\n",
    "    \n",
    "        # Extract the significant features \n",
    "        significant.features = getSignificantFeatures(km.pvalue.table, pvalue.thresh = 0.05)\n",
    "\n",
    "        # Store \n",
    "        significant.features.ls[[end.point]] = significant.features\n",
    "        \n",
    "    } else {\n",
    "        significant.features.ls[[end.point]] = NULL\n",
    "    }\n",
    "}\n",
    "\n",
    "saveRDS(significant.features.ls, file.path(dir.res.root, \"significant.features.ls\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ab79640-0cba-4199-a1f0-f9a28180d0c7",
   "metadata": {},
   "source": [
    "## Feature selection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1a446bc-a03b-46a7-a6bb-7136230db16f",
   "metadata": {},
   "source": [
    "Here we experiment with random survival forests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "05c1b29f-85d4-4dac-b2dd-b863d5b5ece0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      " randomForestSRC 3.2.2 \n",
      " \n",
      " Type rfsrc.news() to see new features, changes, and bug fixes. \n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘randomForestSRC’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    partial\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(pec)\n",
    "library(randomForestSRC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e5383b99-2441-429c-a732-dcae9add9c0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Function for tuning model parameters\n",
    "#\n",
    "#\n",
    "\n",
    "tune_model <- function(data, formula, ntree, seed) {    \n",
    "  \n",
    "  set.seed(seed)\n",
    "  tuning <- tune(formula, data, mtryStart = 2, ntreeTry = ntree, doBest = TRUE)\n",
    "  node <- tuning$optimal[1]\n",
    "  mtry <- tuning$optimal[2]\n",
    "  \n",
    "  error <- tuning$rf$err.rate[500]\n",
    "  res <- tibble(nodesize = node, mtry = mtry, OOB_error = error)\n",
    "  \n",
    "  return(res)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "00060a96-5df7-470f-a384-4dc26d274209",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# Main function for generating the RSF model \n",
    "# based on training set\n",
    "#\n",
    "generateRSFModel = function(training.data,\n",
    "                            variables,\n",
    "                            end.point.event,\n",
    "                            end.point.time, \n",
    "                            model.name,\n",
    "                            ntrees = c(50,100,200,500,1000),\n",
    "                            seed) {\n",
    "  \n",
    "  \n",
    "  # Store relevant data into an object \n",
    "  rsf.obj = list(\"model\" = model.name)\n",
    "\n",
    "  # Select only variables of interest \n",
    "  training.data = training.data  %>% dplyr::select(all_of(c(variables, end.point.event, end.point.time)))\n",
    "\n",
    "  # Generate the formula for the model \n",
    "  surv.expression = paste0(\"Surv(\", end.point.time, \", \" , end.point.event, \")\")\n",
    "  f <- as.formula(paste(surv.expression, \".\", sep = \" ~ \"))\n",
    "\n",
    "  # Store formula\n",
    "  rsf.obj$formula = f\n",
    " \n",
    "  # First tune the parameters\n",
    "  tune.res <- list()\n",
    "  for(n in ntrees) {\n",
    "    tune.res[[as.character(n)]] <- tune_model(training.data, f, n, seed) # Store optimal values for each ntree\n",
    "  }\n",
    "  \n",
    "  # Generate a table \n",
    "  tune.res.df = dplyr::bind_rows(tune.res, .id = \"ntree\")\n",
    "  \n",
    "  # Store tuning results \n",
    "  rsf.obj$tuning = tune.res.df\n",
    "  \n",
    "  idx.min <- which.min(sapply(tune.res, function(x) x$OOB_error)) # Get ntree with smallest error\n",
    "  ntree.optm <- ntrees[idx.min]\n",
    "  nodesize.optm <- tune.res[[idx.min]]$nodesize # Get the best nodesize for the chosen ntree\n",
    "  mtry.optm <- tune.res[[idx.min]]$mtry # Get the best mtry for the chosen ntree\n",
    "  \n",
    "  # Finally fit the model with optimised parameters\n",
    "  set.seed(seed)\n",
    "  rf.fit.train <- list(eval(bquote (rfsrc(.(f), training.data, \n",
    "                                          ntree = ntree.optm, \n",
    "                                          mtry = mtry.optm, \n",
    "                                          nodesize = nodesize.optm, \n",
    "                                          importance = TRUE, \n",
    "                                          block.size = 1, \n",
    "                                          seed = -13))))\n",
    "  \n",
    "  # Store the fitted model\n",
    "  rsf.obj$fit.model = rf.fit.train\n",
    "  \n",
    "  return(rsf.obj)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c663b6a8-7930-4f5c-a5b6-c2a3c16aaa45",
   "metadata": {},
   "source": [
    "PFI "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1894b6c4-d950-4f93-8efd-10f9b32b017e",
   "metadata": {},
   "source": [
    "Iterative feature elimation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "fae18e2c-0835-47f5-ba21-cbb7abd64107",
   "metadata": {},
   "outputs": [],
   "source": [
    "iterativeFeatElim = function(data, \n",
    "                             variables, \n",
    "                             end.point.event,\n",
    "                             end.point.time,\n",
    "                             model.name = \"model2\",\n",
    "                             drop.out.perc = 0.2,\n",
    "                             ntrees = c(50,100,200,500,1000),\n",
    "                             seed){\n",
    "\n",
    "\n",
    "  # Store results for testing \n",
    "  vimp.result.ls = list() \n",
    "\n",
    "  # Errors \n",
    "  oob.errors = list()\n",
    "  \n",
    "  # Num variables\n",
    "  num.variables = list()\n",
    "  \n",
    "  # Store optimal parameters \n",
    "  params.ls = list(\"ntree\" = list(),\n",
    "                   \"mtry\" = list(),\n",
    "                   \"nodesize\" = list())\n",
    "  \n",
    "  # Initialise the number of variables\n",
    "  variables.left = variables[-c(1,2)]\n",
    "  \n",
    "  # Iteration\n",
    "  i = 1\n",
    "    \n",
    "  # If the number of variables is less than 0.2 of the original \n",
    "  # data stop iteration\n",
    "  n.var.stop = round(length(variables) * drop.out.perc)\n",
    "  \n",
    "  # Iterate \n",
    "  while (length(variables.left) > n.var.stop) {\n",
    "   \n",
    "    print(paste0(\"Iteration : \", i))\n",
    "\n",
    "    # Create model \n",
    "    rsf.model = generateRSFModel(data,\n",
    "                 variables = variables.left,\n",
    "                 end.point.event = end.point.event,\n",
    "                 end.point.time = end.point.time,\n",
    "                 model.name = \"model2\",\n",
    "                 ntrees = ntrees,\n",
    "                 seed = seed)\n",
    "    \n",
    "    # Store the params \n",
    "    params.ls[[\"ntree\"]][i] = rsf.model$fit.model[[1]]$ntree\n",
    "    params.ls[[\"mtry\"]][i] = rsf.model$fit.model[[1]]$mtry\n",
    "    params.ls[[\"nodesize\"]][i] = rsf.model$fit.model[[1]]$nodesize\n",
    "  \n",
    "    # Add the oob.error to list   \n",
    "    oob.errors[[i]] = rsf.model$fit.model[[1]]$err.rate[rsf.model$fit.model[[1]]$ntree]\n",
    "  \n",
    "    # Add the number of variables \n",
    "    num.variables[[i]] = length(variables.left)\n",
    "  \n",
    "    # Calculate VIMP\n",
    "    vimp.res = vimp(rsf.model$fit.model[[1]])\n",
    "    \n",
    "    # Prepare a table\n",
    "    vimp.table = as.data.frame(vimp.res$importance) %>% \n",
    "                      tibble::rownames_to_column(\"Variable\") \n",
    "    colnames(vimp.table) = c(\"Variable\", \"Importance\")\n",
    "    \n",
    "    # Order by decreasing Importance such that the worst features\n",
    "    # go bottom of the table\n",
    "    vimp.table = dplyr::arrange(vimp.table, desc(Importance))\n",
    "\n",
    "    # Store results \n",
    "    vimp.result.ls[[i]] = vimp.table\n",
    "    \n",
    "    # Remove the worst performing features \n",
    "    variables.left = vimp.table$Variable[1:((1 - drop.out.perc) * length(variables.left))]\n",
    "    \n",
    "    # Increment \n",
    "    i = i + 1\n",
    "  }\n",
    "  \n",
    "  # Combine to table\n",
    "  final.df = data.frame(\"ntree\" = unlist(params.ls$ntree),\n",
    "             \"mtry\" = unlist(params.ls$mtry),\n",
    "             \"nodesize\" = unlist(params.ls$nodesize),\n",
    "             \"num_variable\" = unlist(num.variables),\n",
    "             \"error\" = unlist(oob.errors))\n",
    "\n",
    "\n",
    "  \n",
    "  # Find row with minumum error \n",
    "  min.idx = which(final.df$error == min(final.df$error))\n",
    "  print(min.idx)\n",
    "  \n",
    "  # Select the last row having error equal to minimum. We want to minimize the \n",
    "  # number of features\n",
    "  if (length(min.idx) > 1) {\n",
    "    min.idx = max(min.idx)\n",
    "  }\n",
    "  print(min.idx)\n",
    "  \n",
    "  # Optimal parameters \n",
    "  optimal.set.param = final.df[min.idx,]\n",
    "  optimal.set.variables = vimp.result.ls[[min.idx]]\n",
    "  \n",
    "  \n",
    "  # Final results\n",
    "  final.results = list(\"Errors\" = final.df, \n",
    "                       \"Features\" = vimp.result.ls,\n",
    "                       \"Optimal.param\" = optimal.set.param,\n",
    "                       \"Optimal.var\" = optimal.set.variables)\n",
    "  \n",
    "  return(final.results)\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7a0c275-5d6e-4000-8245-72c76e199405",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b66392bf-801a-458a-8645-4b50e6820b4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "end.point.time = \"PFI.time.clin\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "deea6feb-48aa-4974-942b-6b561e7c11a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature.elim.pfi = iterativeFeatElim(data = tcga.dataset.merged$PFI$train,\n",
    "                  variables = variables.selected.filtered.ls[[\"PFI\"]],\n",
    "                  end.point.event = \"PFI.clin\",\n",
    "                  end.point.time = \"PFI.time.clin\",\n",
    "                  model.name = \"model2\",\n",
    "                  drop.out.perc = 0.2,\n",
    "                  ntrees = c(50,100,200,500,1000),\n",
    "                  seed = 42)    \n",
    "                  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "b6e4ce75-a128-4480-8267-9fe8b8a4cf29",
   "metadata": {},
   "outputs": [],
   "source": [
    "feat.elim.res = readRDS(file.path(out.dir.data, \"Feature_elimination_res_PFI.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "22433bdf-0f6e-4f0b-a256-13dff01c1b21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 7 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>ntree</th><th scope=col>mtry</th><th scope=col>nodesize</th><th scope=col>num_variable</th><th scope=col>error</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>100</td><td>1</td><td> 8</td><td>35</td><td>0.3309813</td></tr>\n",
       "\t<tr><td>500</td><td>1</td><td>25</td><td>28</td><td>0.3030503</td></tr>\n",
       "\t<tr><td> 50</td><td>7</td><td> 8</td><td>22</td><td>0.3006983</td></tr>\n",
       "\t<tr><td>100</td><td>1</td><td> 8</td><td>17</td><td>0.2995222</td></tr>\n",
       "\t<tr><td>500</td><td>4</td><td> 2</td><td>13</td><td>0.3409776</td></tr>\n",
       "\t<tr><td>500</td><td>5</td><td> 1</td><td>10</td><td>0.3189269</td></tr>\n",
       "\t<tr><td> 50</td><td>6</td><td> 2</td><td> 8</td><td>0.3216465</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 7 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " ntree & mtry & nodesize & num\\_variable & error\\\\\n",
       " <dbl> & <dbl> & <dbl> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 100 & 1 &  8 & 35 & 0.3309813\\\\\n",
       "\t 500 & 1 & 25 & 28 & 0.3030503\\\\\n",
       "\t  50 & 7 &  8 & 22 & 0.3006983\\\\\n",
       "\t 100 & 1 &  8 & 17 & 0.2995222\\\\\n",
       "\t 500 & 4 &  2 & 13 & 0.3409776\\\\\n",
       "\t 500 & 5 &  1 & 10 & 0.3189269\\\\\n",
       "\t  50 & 6 &  2 &  8 & 0.3216465\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 7 × 5\n",
       "\n",
       "| ntree &lt;dbl&gt; | mtry &lt;dbl&gt; | nodesize &lt;dbl&gt; | num_variable &lt;int&gt; | error &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 100 | 1 |  8 | 35 | 0.3309813 |\n",
       "| 500 | 1 | 25 | 28 | 0.3030503 |\n",
       "|  50 | 7 |  8 | 22 | 0.3006983 |\n",
       "| 100 | 1 |  8 | 17 | 0.2995222 |\n",
       "| 500 | 4 |  2 | 13 | 0.3409776 |\n",
       "| 500 | 5 |  1 | 10 | 0.3189269 |\n",
       "|  50 | 6 |  2 |  8 | 0.3216465 |\n",
       "\n"
      ],
      "text/plain": [
       "  ntree mtry nodesize num_variable error    \n",
       "1 100   1     8       35           0.3309813\n",
       "2 500   1    25       28           0.3030503\n",
       "3  50   7     8       22           0.3006983\n",
       "4 100   1     8       17           0.2995222\n",
       "5 500   4     2       13           0.3409776\n",
       "6 500   5     1       10           0.3189269\n",
       "7  50   6     2        8           0.3216465"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "feat.elim.res$Errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "18d8d919-6b61-4c2a-b4a8-96543333d862",
   "metadata": {},
   "outputs": [],
   "source": [
    "vip.collected = bind_rows(feat.elim.res$Features, .id = \"Iteration\") %>% pivot_wider(names_from = Iteration, values_from = Importance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "0190423d-e81c-4f8c-884e-b253d57a3955",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 35 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Variable</th><th scope=col>1</th><th scope=col>2</th><th scope=col>3</th><th scope=col>4</th><th scope=col>5</th><th scope=col>6</th><th scope=col>7</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>AURKB.exp   </td><td> 0.0112466516</td><td>0.024319146</td><td> 0.025517688</td><td> 0.0166860983</td><td>0.041220932</td><td>0.070169715</td><td>0.10987642</td></tr>\n",
       "\t<tr><td>PTEN.exp    </td><td> 0.0153373213</td><td>0.027285531</td><td> 0.022245246</td><td> 0.0315808193</td><td>0.035677535</td><td>0.061739104</td><td>0.09847870</td></tr>\n",
       "\t<tr><td>TGFB1.exp   </td><td> 0.0168543497</td><td>0.016618490</td><td> 0.032298196</td><td> 0.0235684475</td><td>0.028683854</td><td>0.036847030</td><td>0.07719605</td></tr>\n",
       "\t<tr><td>KIF2C.exp   </td><td> 0.0078198308</td><td>0.025926463</td><td> 0.047219532</td><td> 0.0522887796</td><td>0.025678605</td><td>0.020767503</td><td>0.05683456</td></tr>\n",
       "\t<tr><td>BIRC5.exp   </td><td> 0.0075572326</td><td>0.016653425</td><td> 0.040349042</td><td> 0.0116136525</td><td>0.016075173</td><td>0.018383435</td><td>0.04026610</td></tr>\n",
       "\t<tr><td>GDF9.exp    </td><td> 0.0057092189</td><td>0.010278202</td><td> 0.015200859</td><td> 0.0309803598</td><td>0.012100446</td><td>0.047000375</td><td>0.03673613</td></tr>\n",
       "\t<tr><td>FLCN.exp    </td><td> 0.0101900738</td><td>0.023873504</td><td> 0.032028799</td><td> 0.0398617795</td><td>0.031163148</td><td>0.029970969</td><td>0.02960134</td></tr>\n",
       "\t<tr><td>ENG.exp     </td><td> 0.0087708146</td><td>0.012458479</td><td> 0.028244611</td><td> 0.0313942057</td><td>0.010719336</td><td>0.020268861</td><td>0.01514332</td></tr>\n",
       "\t<tr><td>CDCA8.exp   </td><td> 0.0084615116</td><td>0.012342167</td><td> 0.020615266</td><td> 0.0182926493</td><td>0.014705465</td><td>0.011720581</td><td>        NA</td></tr>\n",
       "\t<tr><td>SMAD9.exp   </td><td> 0.0058374508</td><td>0.009179841</td><td> 0.019560365</td><td> 0.0233721015</td><td>0.009771761</td><td>0.003302699</td><td>        NA</td></tr>\n",
       "\t<tr><td>BMPR1A.exp  </td><td> 0.0064310692</td><td>0.007110557</td><td> 0.031261730</td><td> 0.0133612447</td><td>0.007109471</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>PSEN1.exp   </td><td> 0.0035497569</td><td>0.011618814</td><td> 0.042877197</td><td> 0.0226689974</td><td>0.006298239</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>WFIKKN1.exp </td><td> 0.0164100465</td><td>0.013035348</td><td> 0.016106154</td><td> 0.0203134480</td><td>0.006052286</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>KIF23.exp   </td><td> 0.0206806097</td><td>0.021428039</td><td> 0.030372107</td><td> 0.0097231981</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>USP15.exp   </td><td> 0.0104182179</td><td>0.007857814</td><td> 0.011251504</td><td> 0.0074119165</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>ITGB8.exp   </td><td> 0.0117412476</td><td>0.006877598</td><td> 0.010960953</td><td> 0.0009402809</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>ID1.exp     </td><td> 0.0109016476</td><td>0.015197496</td><td> 0.010575668</td><td>-0.0066405445</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>VPS4B.exp   </td><td> 0.0081490338</td><td>0.008256080</td><td> 0.004936084</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>GLG1.exp    </td><td> 0.0066284737</td><td>0.010294664</td><td> 0.004295977</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>AURKA.exp   </td><td> 0.0130804666</td><td>0.022553716</td><td> 0.003031191</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>HIPK2.exp   </td><td> 0.0049017959</td><td>0.007057320</td><td>-0.001736990</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>SMAD4.exp   </td><td> 0.0091629309</td><td>0.014090893</td><td>-0.002001280</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>ADAM9.exp   </td><td> 0.0104277191</td><td>0.004809119</td><td>          NA</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>TP53.cn     </td><td> 0.0104078767</td><td>0.006314664</td><td>          NA</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>ACVRL1.exp  </td><td> 0.0045484202</td><td>0.006303315</td><td>          NA</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>GDF10.exp   </td><td> 0.0032801471</td><td>0.005986634</td><td>          NA</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>NDP.exp     </td><td>-0.0003169984</td><td>0.003601148</td><td>          NA</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>TGFBRAP1.exp</td><td>-0.0003947136</td><td>0.004827292</td><td>          NA</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>SIRT1.exp   </td><td>-0.0004386437</td><td>         NA</td><td>          NA</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>PXN.exp     </td><td>-0.0019791322</td><td>         NA</td><td>          NA</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>ERO1A.exp   </td><td>-0.0035632564</td><td>         NA</td><td>          NA</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>SMAD2.exp   </td><td>-0.0044896117</td><td>         NA</td><td>          NA</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>NRROS.exp   </td><td>-0.0082528151</td><td>         NA</td><td>          NA</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>ARRB2.exp   </td><td>-0.0108260640</td><td>         NA</td><td>          NA</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "\t<tr><td>LTBP3.exp   </td><td>-0.0149560607</td><td>         NA</td><td>          NA</td><td>           NA</td><td>         NA</td><td>         NA</td><td>        NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 35 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " Variable & 1 & 2 & 3 & 4 & 5 & 6 & 7\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t AURKB.exp    &  0.0112466516 & 0.024319146 &  0.025517688 &  0.0166860983 & 0.041220932 & 0.070169715 & 0.10987642\\\\\n",
       "\t PTEN.exp     &  0.0153373213 & 0.027285531 &  0.022245246 &  0.0315808193 & 0.035677535 & 0.061739104 & 0.09847870\\\\\n",
       "\t TGFB1.exp    &  0.0168543497 & 0.016618490 &  0.032298196 &  0.0235684475 & 0.028683854 & 0.036847030 & 0.07719605\\\\\n",
       "\t KIF2C.exp    &  0.0078198308 & 0.025926463 &  0.047219532 &  0.0522887796 & 0.025678605 & 0.020767503 & 0.05683456\\\\\n",
       "\t BIRC5.exp    &  0.0075572326 & 0.016653425 &  0.040349042 &  0.0116136525 & 0.016075173 & 0.018383435 & 0.04026610\\\\\n",
       "\t GDF9.exp     &  0.0057092189 & 0.010278202 &  0.015200859 &  0.0309803598 & 0.012100446 & 0.047000375 & 0.03673613\\\\\n",
       "\t FLCN.exp     &  0.0101900738 & 0.023873504 &  0.032028799 &  0.0398617795 & 0.031163148 & 0.029970969 & 0.02960134\\\\\n",
       "\t ENG.exp      &  0.0087708146 & 0.012458479 &  0.028244611 &  0.0313942057 & 0.010719336 & 0.020268861 & 0.01514332\\\\\n",
       "\t CDCA8.exp    &  0.0084615116 & 0.012342167 &  0.020615266 &  0.0182926493 & 0.014705465 & 0.011720581 &         NA\\\\\n",
       "\t SMAD9.exp    &  0.0058374508 & 0.009179841 &  0.019560365 &  0.0233721015 & 0.009771761 & 0.003302699 &         NA\\\\\n",
       "\t BMPR1A.exp   &  0.0064310692 & 0.007110557 &  0.031261730 &  0.0133612447 & 0.007109471 &          NA &         NA\\\\\n",
       "\t PSEN1.exp    &  0.0035497569 & 0.011618814 &  0.042877197 &  0.0226689974 & 0.006298239 &          NA &         NA\\\\\n",
       "\t WFIKKN1.exp  &  0.0164100465 & 0.013035348 &  0.016106154 &  0.0203134480 & 0.006052286 &          NA &         NA\\\\\n",
       "\t KIF23.exp    &  0.0206806097 & 0.021428039 &  0.030372107 &  0.0097231981 &          NA &          NA &         NA\\\\\n",
       "\t USP15.exp    &  0.0104182179 & 0.007857814 &  0.011251504 &  0.0074119165 &          NA &          NA &         NA\\\\\n",
       "\t ITGB8.exp    &  0.0117412476 & 0.006877598 &  0.010960953 &  0.0009402809 &          NA &          NA &         NA\\\\\n",
       "\t ID1.exp      &  0.0109016476 & 0.015197496 &  0.010575668 & -0.0066405445 &          NA &          NA &         NA\\\\\n",
       "\t VPS4B.exp    &  0.0081490338 & 0.008256080 &  0.004936084 &            NA &          NA &          NA &         NA\\\\\n",
       "\t GLG1.exp     &  0.0066284737 & 0.010294664 &  0.004295977 &            NA &          NA &          NA &         NA\\\\\n",
       "\t AURKA.exp    &  0.0130804666 & 0.022553716 &  0.003031191 &            NA &          NA &          NA &         NA\\\\\n",
       "\t HIPK2.exp    &  0.0049017959 & 0.007057320 & -0.001736990 &            NA &          NA &          NA &         NA\\\\\n",
       "\t SMAD4.exp    &  0.0091629309 & 0.014090893 & -0.002001280 &            NA &          NA &          NA &         NA\\\\\n",
       "\t ADAM9.exp    &  0.0104277191 & 0.004809119 &           NA &            NA &          NA &          NA &         NA\\\\\n",
       "\t TP53.cn      &  0.0104078767 & 0.006314664 &           NA &            NA &          NA &          NA &         NA\\\\\n",
       "\t ACVRL1.exp   &  0.0045484202 & 0.006303315 &           NA &            NA &          NA &          NA &         NA\\\\\n",
       "\t GDF10.exp    &  0.0032801471 & 0.005986634 &           NA &            NA &          NA &          NA &         NA\\\\\n",
       "\t NDP.exp      & -0.0003169984 & 0.003601148 &           NA &            NA &          NA &          NA &         NA\\\\\n",
       "\t TGFBRAP1.exp & -0.0003947136 & 0.004827292 &           NA &            NA &          NA &          NA &         NA\\\\\n",
       "\t SIRT1.exp    & -0.0004386437 &          NA &           NA &            NA &          NA &          NA &         NA\\\\\n",
       "\t PXN.exp      & -0.0019791322 &          NA &           NA &            NA &          NA &          NA &         NA\\\\\n",
       "\t ERO1A.exp    & -0.0035632564 &          NA &           NA &            NA &          NA &          NA &         NA\\\\\n",
       "\t SMAD2.exp    & -0.0044896117 &          NA &           NA &            NA &          NA &          NA &         NA\\\\\n",
       "\t NRROS.exp    & -0.0082528151 &          NA &           NA &            NA &          NA &          NA &         NA\\\\\n",
       "\t ARRB2.exp    & -0.0108260640 &          NA &           NA &            NA &          NA &          NA &         NA\\\\\n",
       "\t LTBP3.exp    & -0.0149560607 &          NA &           NA &            NA &          NA &          NA &         NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 35 × 8\n",
       "\n",
       "| Variable &lt;chr&gt; | 1 &lt;dbl&gt; | 2 &lt;dbl&gt; | 3 &lt;dbl&gt; | 4 &lt;dbl&gt; | 5 &lt;dbl&gt; | 6 &lt;dbl&gt; | 7 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| AURKB.exp    |  0.0112466516 | 0.024319146 |  0.025517688 |  0.0166860983 | 0.041220932 | 0.070169715 | 0.10987642 |\n",
       "| PTEN.exp     |  0.0153373213 | 0.027285531 |  0.022245246 |  0.0315808193 | 0.035677535 | 0.061739104 | 0.09847870 |\n",
       "| TGFB1.exp    |  0.0168543497 | 0.016618490 |  0.032298196 |  0.0235684475 | 0.028683854 | 0.036847030 | 0.07719605 |\n",
       "| KIF2C.exp    |  0.0078198308 | 0.025926463 |  0.047219532 |  0.0522887796 | 0.025678605 | 0.020767503 | 0.05683456 |\n",
       "| BIRC5.exp    |  0.0075572326 | 0.016653425 |  0.040349042 |  0.0116136525 | 0.016075173 | 0.018383435 | 0.04026610 |\n",
       "| GDF9.exp     |  0.0057092189 | 0.010278202 |  0.015200859 |  0.0309803598 | 0.012100446 | 0.047000375 | 0.03673613 |\n",
       "| FLCN.exp     |  0.0101900738 | 0.023873504 |  0.032028799 |  0.0398617795 | 0.031163148 | 0.029970969 | 0.02960134 |\n",
       "| ENG.exp      |  0.0087708146 | 0.012458479 |  0.028244611 |  0.0313942057 | 0.010719336 | 0.020268861 | 0.01514332 |\n",
       "| CDCA8.exp    |  0.0084615116 | 0.012342167 |  0.020615266 |  0.0182926493 | 0.014705465 | 0.011720581 |         NA |\n",
       "| SMAD9.exp    |  0.0058374508 | 0.009179841 |  0.019560365 |  0.0233721015 | 0.009771761 | 0.003302699 |         NA |\n",
       "| BMPR1A.exp   |  0.0064310692 | 0.007110557 |  0.031261730 |  0.0133612447 | 0.007109471 |          NA |         NA |\n",
       "| PSEN1.exp    |  0.0035497569 | 0.011618814 |  0.042877197 |  0.0226689974 | 0.006298239 |          NA |         NA |\n",
       "| WFIKKN1.exp  |  0.0164100465 | 0.013035348 |  0.016106154 |  0.0203134480 | 0.006052286 |          NA |         NA |\n",
       "| KIF23.exp    |  0.0206806097 | 0.021428039 |  0.030372107 |  0.0097231981 |          NA |          NA |         NA |\n",
       "| USP15.exp    |  0.0104182179 | 0.007857814 |  0.011251504 |  0.0074119165 |          NA |          NA |         NA |\n",
       "| ITGB8.exp    |  0.0117412476 | 0.006877598 |  0.010960953 |  0.0009402809 |          NA |          NA |         NA |\n",
       "| ID1.exp      |  0.0109016476 | 0.015197496 |  0.010575668 | -0.0066405445 |          NA |          NA |         NA |\n",
       "| VPS4B.exp    |  0.0081490338 | 0.008256080 |  0.004936084 |            NA |          NA |          NA |         NA |\n",
       "| GLG1.exp     |  0.0066284737 | 0.010294664 |  0.004295977 |            NA |          NA |          NA |         NA |\n",
       "| AURKA.exp    |  0.0130804666 | 0.022553716 |  0.003031191 |            NA |          NA |          NA |         NA |\n",
       "| HIPK2.exp    |  0.0049017959 | 0.007057320 | -0.001736990 |            NA |          NA |          NA |         NA |\n",
       "| SMAD4.exp    |  0.0091629309 | 0.014090893 | -0.002001280 |            NA |          NA |          NA |         NA |\n",
       "| ADAM9.exp    |  0.0104277191 | 0.004809119 |           NA |            NA |          NA |          NA |         NA |\n",
       "| TP53.cn      |  0.0104078767 | 0.006314664 |           NA |            NA |          NA |          NA |         NA |\n",
       "| ACVRL1.exp   |  0.0045484202 | 0.006303315 |           NA |            NA |          NA |          NA |         NA |\n",
       "| GDF10.exp    |  0.0032801471 | 0.005986634 |           NA |            NA |          NA |          NA |         NA |\n",
       "| NDP.exp      | -0.0003169984 | 0.003601148 |           NA |            NA |          NA |          NA |         NA |\n",
       "| TGFBRAP1.exp | -0.0003947136 | 0.004827292 |           NA |            NA |          NA |          NA |         NA |\n",
       "| SIRT1.exp    | -0.0004386437 |          NA |           NA |            NA |          NA |          NA |         NA |\n",
       "| PXN.exp      | -0.0019791322 |          NA |           NA |            NA |          NA |          NA |         NA |\n",
       "| ERO1A.exp    | -0.0035632564 |          NA |           NA |            NA |          NA |          NA |         NA |\n",
       "| SMAD2.exp    | -0.0044896117 |          NA |           NA |            NA |          NA |          NA |         NA |\n",
       "| NRROS.exp    | -0.0082528151 |          NA |           NA |            NA |          NA |          NA |         NA |\n",
       "| ARRB2.exp    | -0.0108260640 |          NA |           NA |            NA |          NA |          NA |         NA |\n",
       "| LTBP3.exp    | -0.0149560607 |          NA |           NA |            NA |          NA |          NA |         NA |\n",
       "\n"
      ],
      "text/plain": [
       "   Variable     1             2           3            4            \n",
       "1  AURKB.exp     0.0112466516 0.024319146  0.025517688  0.0166860983\n",
       "2  PTEN.exp      0.0153373213 0.027285531  0.022245246  0.0315808193\n",
       "3  TGFB1.exp     0.0168543497 0.016618490  0.032298196  0.0235684475\n",
       "4  KIF2C.exp     0.0078198308 0.025926463  0.047219532  0.0522887796\n",
       "5  BIRC5.exp     0.0075572326 0.016653425  0.040349042  0.0116136525\n",
       "6  GDF9.exp      0.0057092189 0.010278202  0.015200859  0.0309803598\n",
       "7  FLCN.exp      0.0101900738 0.023873504  0.032028799  0.0398617795\n",
       "8  ENG.exp       0.0087708146 0.012458479  0.028244611  0.0313942057\n",
       "9  CDCA8.exp     0.0084615116 0.012342167  0.020615266  0.0182926493\n",
       "10 SMAD9.exp     0.0058374508 0.009179841  0.019560365  0.0233721015\n",
       "11 BMPR1A.exp    0.0064310692 0.007110557  0.031261730  0.0133612447\n",
       "12 PSEN1.exp     0.0035497569 0.011618814  0.042877197  0.0226689974\n",
       "13 WFIKKN1.exp   0.0164100465 0.013035348  0.016106154  0.0203134480\n",
       "14 KIF23.exp     0.0206806097 0.021428039  0.030372107  0.0097231981\n",
       "15 USP15.exp     0.0104182179 0.007857814  0.011251504  0.0074119165\n",
       "16 ITGB8.exp     0.0117412476 0.006877598  0.010960953  0.0009402809\n",
       "17 ID1.exp       0.0109016476 0.015197496  0.010575668 -0.0066405445\n",
       "18 VPS4B.exp     0.0081490338 0.008256080  0.004936084            NA\n",
       "19 GLG1.exp      0.0066284737 0.010294664  0.004295977            NA\n",
       "20 AURKA.exp     0.0130804666 0.022553716  0.003031191            NA\n",
       "21 HIPK2.exp     0.0049017959 0.007057320 -0.001736990            NA\n",
       "22 SMAD4.exp     0.0091629309 0.014090893 -0.002001280            NA\n",
       "23 ADAM9.exp     0.0104277191 0.004809119           NA            NA\n",
       "24 TP53.cn       0.0104078767 0.006314664           NA            NA\n",
       "25 ACVRL1.exp    0.0045484202 0.006303315           NA            NA\n",
       "26 GDF10.exp     0.0032801471 0.005986634           NA            NA\n",
       "27 NDP.exp      -0.0003169984 0.003601148           NA            NA\n",
       "28 TGFBRAP1.exp -0.0003947136 0.004827292           NA            NA\n",
       "29 SIRT1.exp    -0.0004386437          NA           NA            NA\n",
       "30 PXN.exp      -0.0019791322          NA           NA            NA\n",
       "31 ERO1A.exp    -0.0035632564          NA           NA            NA\n",
       "32 SMAD2.exp    -0.0044896117          NA           NA            NA\n",
       "33 NRROS.exp    -0.0082528151          NA           NA            NA\n",
       "34 ARRB2.exp    -0.0108260640          NA           NA            NA\n",
       "35 LTBP3.exp    -0.0149560607          NA           NA            NA\n",
       "   5           6           7         \n",
       "1  0.041220932 0.070169715 0.10987642\n",
       "2  0.035677535 0.061739104 0.09847870\n",
       "3  0.028683854 0.036847030 0.07719605\n",
       "4  0.025678605 0.020767503 0.05683456\n",
       "5  0.016075173 0.018383435 0.04026610\n",
       "6  0.012100446 0.047000375 0.03673613\n",
       "7  0.031163148 0.029970969 0.02960134\n",
       "8  0.010719336 0.020268861 0.01514332\n",
       "9  0.014705465 0.011720581         NA\n",
       "10 0.009771761 0.003302699         NA\n",
       "11 0.007109471          NA         NA\n",
       "12 0.006298239          NA         NA\n",
       "13 0.006052286          NA         NA\n",
       "14          NA          NA         NA\n",
       "15          NA          NA         NA\n",
       "16          NA          NA         NA\n",
       "17          NA          NA         NA\n",
       "18          NA          NA         NA\n",
       "19          NA          NA         NA\n",
       "20          NA          NA         NA\n",
       "21          NA          NA         NA\n",
       "22          NA          NA         NA\n",
       "23          NA          NA         NA\n",
       "24          NA          NA         NA\n",
       "25          NA          NA         NA\n",
       "26          NA          NA         NA\n",
       "27          NA          NA         NA\n",
       "28          NA          NA         NA\n",
       "29          NA          NA         NA\n",
       "30          NA          NA         NA\n",
       "31          NA          NA         NA\n",
       "32          NA          NA         NA\n",
       "33          NA          NA         NA\n",
       "34          NA          NA         NA\n",
       "35          NA          NA         NA"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "vip.collected %>% arrange(desc(`7`), desc(`6`), desc(`5`), desc(`4`), desc(`3`))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ded4fbff-8a6e-481d-b6af-9eab2fff5c19",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
